qrexec-lib/validator-test.c

```diff
@@ -70,7 +72,137 @@ int main(int argc, char **argv)
+    assert(!qubes_pure_validate_file_name((const uint8_t *)"../a"));
+    assert(!qubes_pure_validate_file_name((const uint8_t *)"a/../a"));
+
+    // Looks like "." or ".." but is not
+    assert(qubes_pure_validate_file_name((const uint8_t *)".a"));
+    assert(qubes_pure_validate_file_name((const uint8_t *)"..a"));
+
+    // Charset validation checks
+    for (unsigned int i = 0; i < 256; ++i) {
+        // These bytes are forbidden even if Unicode filtering is off.
+        bool always_forbidden = i == '/' || i == '.' || i == 0;
+        // NUL is not a valid continuation byte, so if Unicode filtering
+        // is on, all non-ASCII bytes are forbidden.
+        bool bad_unicode = i < 0x20 || i > 0x7E;
+        unsigned char buf[2] = { i, 0 };
+        assert(qubes_pure_validate_file_name_v2(buf, QUBES_PURE_ALLOW_UNSAFE_CHARACTERS) ==
+               (always_forbidden ? -EILSEQ : 0));
+        assert(qubes_pure_validate_file_name_v2(buf, 0) ==
+               (always_forbidden || bad_unicode ? -EILSEQ : 0));
+    }
+
+    struct p {
+        const char *const path, *const target, *const file;
+        int const line, flags;
+        bool const allowed;
+    } symlink_checks[] = {
+#define TEST(path_, target_, flags_, allowed_) \
+        { .path = (path_)                      \
+        , .target = (target_)                  \
+        , .file = __FILE__                     \
+        , .line = __LINE__                     \
+        , .flags = (flags_)                    \
+        , .allowed = (allowed_)                \
+        }
+        // Symbolic links
+        // Top level cannot be symlink
+        TEST("a", "b", 0, false),
+        TEST("a/b", "../a", 0, false),
+        TEST("a", "b", 0, false),
+        // Symbolic links cannot escape
+        TEST("a/b", "../a", 0, false),
+        TEST("a/b", "../a", 0, false),
+        TEST("a/b", "../a/b/c", 0, false),
+        TEST("a/b", "../a/b/c", 0, false),
+        TEST("a/b/c", "../../a", 0, false),
+        TEST("a/b/c", "../../a", 0, false),
+        TEST("a/b", "a", 0, true),
+        TEST("a/b/c", "../a", 0, true),
+        // Absolute symlinks are rejected
+        TEST("a/b/c", "/a", 0, false),
+        TEST("a/b/c", "/a", 0, false),
+        // Symlinks may end in "..".
+        TEST("a/b/c", "..", 0, true),
+        // Symlinks may end in "/".
+        TEST("a/b/c", "a/", 0, true),
+        // Invalid paths are rejected...
+        TEST("..", "a/", 0, false),
+        // ...even with QUBES_PURE_ALLOW_UNSAFE_SYMLINKS.
+        TEST("..", "a/", QUBES_PURE_ALLOW_UNSAFE_SYMLINKS, false),
+        // but QUBES_PURE_ALLOW_UNSAFE_SYMLINKS allows bad symlinks
+        TEST("a", "/bin/sh", QUBES_PURE_ALLOW_UNSAFE_SYMLINKS, true),
+        // QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS allows non-canonical symlinks...
+        TEST("a/b", "b//c", QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS, true),
+        TEST("a/b", "b/./c", QUBES_PURE_ALLOW_NON_CANONICAL_SYMLINKS, true),
```

Comment:
If you check if stuff like this is allowed with the flag, shouldn't you check also if it's refused without it?